### ISSUE TITLE
Procedural macro support

### ISSUE THREAD

**Issue by Undin** [intellij-rust/intellij-rust#6908]
Tracking issue about procedural macro support.

Procedural macro support feature depends on the following experimental features:
- `org.rust.cargo.evaluate.build.scripts` - enables building and collecting build artefacts including proc-macro libraries during importing of project structure
- `org.rust.macros.proc` -  enables expansion of procedural macros

To enable an experimental feature, type `Experimental feature` in the dialog of `Help | Find Action` action and enabled the corresponding item. 

- [x] introduce proc macro server to evaluate procedural macro functions #6658
- [x] compile only "host" code of project: build scripts and proc-macro targets when `org.rust.cargo.evaluate.build.scripts` experimental feature is enabled #6735
- [x] expand [function-like](https://doc.rust-lang.org/reference/procedural-macros.html#function-like-procedural-macros) macros #6564
    Fixes #4707
    Fixes #6367
- [ ] expand [custom derive](https://doc.rust-lang.org/reference/procedural-macros.html#derive-macros) macros
    Fixes #1786
    Fixes #5468
    Fixes #5478
    Fixes #5728
    Fixes #5954
    Fixes #6036
    Fixes #6489
    Fixes #6880
- [ ] expand [attribute](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros) macros
    Fixes #5239
    Fixes #5748
    Fixes #5896
    Fixes #5975 (requires changes in the corresponding inspection)
    Fixes #6091
    Fixes #6162 (requires changes in the annotator)
    Fixes #6482
- [ ] enable `org.rust.cargo.evaluate.build.scripts` experimental feature by default
- [ ] enable `org.rust.macros.proc` experimental feature by default

**Comment by webern** [comment]
Does this mean that a subset of derive macro code is currently considered (e.g. for code completion and missing-symbol highlighting?

> take into account all items generated by custom derive macros

For example, the Snafu derive macro produces structs which intellij-rust seems to be unaware of.

Would a minimal use case example of this and tracking issue be of any benefit?

**Comment by Undin** [comment]
@webern 
> Does this mean that a subset of derive macro code is currently considered (e.g. for code completion and missing-symbol highlighting?

Yep. Currently, the plugin takes into account only impl blocks generated by derive proc macro

> Would a minimal use case example of this and tracking issue be of any benefit?

I don't think that we need additional tracking issue only for that case because the current issue already contains a similar item: "take into account all items generated by custom derive macros". But an additional use-case can be useful indeed to check it during the corresponding implementation

**Comment by webern** [comment]
> additional use-case can be useful indeed to check it during the corresponding implementation

Here is the use-case:

Cargo.toml:

```toml
[package]
name = "snafu-jetbrains"
version = "0.1.0"
edition = "2018"

[dependencies]
snafu = "0.6"
```

main.rs

```rust
use snafu::Snafu;

/// The error type for this library.
#[derive(Debug, Snafu)]
enum Error {
    #[snafu(display("This is always an error:: {}", message))]
    AlwaysError { message: String },
}

fn hello_world() -> Result<(), Error> {
    AlwaysError {
        message: "this is a struct created by snafu",
    }
    .fail()
}

fn main() {
    hello_world().unwrap()
}
```

Problem:

![image](https://user-images.githubusercontent.com/6260372/121945000-20e14a00-cd08-11eb-8249-a0e6c728b0e3.png)

Desired behavior: code completion should be available for the struct `AlwaysError`:

![image](https://user-images.githubusercontent.com/6260372/121945213-60a83180-cd08-11eb-9413-077c12de7f80.png)

Desired behavior: context actions should provide the ability to fill in missing struct fields for the `AlwaysError` struct.

![image](https://user-images.githubusercontent.com/6260372/121945297-76b5f200-cd08-11eb-9e4e-e0d1b8f4ab82.png)

Thank you! 🚀

**Comment by Undin** [comment]
@webern Thank you for your example. I've filed a separate issue #7365 with it

**Comment by webern** [comment]
How are we coming along on `take into account all items generated by custom derive macros` and #7365? This is suuuuuch a huge feature for me. Is at the top of the intellij-rust priority list?

Right now I have to open Sublime Text using Rust Analyzer to discover generated structs and functions! Help, I'm lost without my IDE!

**Comment by vlad20012** [comment]
@webern #7725

**Comment by hamza1311** [comment]
What's blocking https://github.com/intellij-rust/intellij-rust/issues/6732? I don't see that issue mentioned here. I'm experiencing the same problem and would love if that is fixed.

**Comment by vlad20012** [comment]
@hamza1311, I suppose it's #7725, but you can use the latest `yew` version from GitHub (not from crates.io) that works well right now.

**Comment by hamza1311** [comment]
@vlad20012, I see. It seems that specific case is fixed but there are other cases where the bug still exists. For example, the following code still gives the same message:  
`lib.rs`:
```rust
use yew::prelude::*;

fn _test() -> Html {
    use_effect(|| {
        || {}
    });

    html! {

    }
}
```
`Cargo.toml`:
```toml
[package]
name = "yew-html-macro-bug-repro"
version = "0.1.0"
edition = "2018"

[dependencies]
yew = { git = "https://github.com/yewstack/yew/" }
```

Screenshot:
![image](https://user-images.githubusercontent.com/47357913/130624277-76f3c444-844c-4ab7-bae5-0d594a181f7a.png)

**Comment by Undin** [comment]
@longfellowone could you, please, post the image here itself or provide another link? I see only 404 with this one

**Comment by hectorj** [comment]
I've noted that for sqlx I need to

- set `SQLX_OFFLINE=true` in my project's root `.env`
- run `cargo sqlx prepare`

for `intellij-rust` to be able to expand `query!` macros. (both documented in https://github.com/launchbadge/sqlx/tree/ce801b933006c422ada464d0663c68dcf6c680d4/sqlx-cli#force-building-in-offline-mode)

Setting `DATABASE_URL` in this `.env` does not work. I suspect the macro expander can't reach the database for some reason.

**Comment by tisonkun** [comment]
I think https://github.com/intellij-rust/intellij-rust/issues/8221 could be part of this tracking issue.

**Comment by aldanor** [comment]
With #7194 and with all experimental features enabled, wonder why code highlighting for functions is different (i.e. half-disabled) when applying even a dummy attribute macro like this one which does nothing and just returns the input back?

```rust
// do nothing, just return the input back
#[proc_macro_attribute]
pub fn dummy(_args: TokenStream, input: TokenStream) -> TokenStream {
    input
}
```

and then:

![image](https://user-images.githubusercontent.com/2418513/144726242-e6c67913-5fbe-4431-b3f4-9be9b78ad53b.png)

Can anything be done from the proc macro authors standpoint so that it would be properly recognized by intellij-rust (or is it a bug of some sort and it should already work)? (e.g. in my particular case, I know that the input to a proc macro must be `ItemFn`)

**Comment by longfellowone** [comment]
> I've noted that for sqlx I need to
> 
> * set `SQLX_OFFLINE=true` in my project's root `.env`
> * run `cargo sqlx prepare`
> 
> for `intellij-rust` to be able to expand `query!` macros. (both documented in https://github.com/launchbadge/sqlx/tree/ce801b933006c422ada464d0663c68dcf6c680d4/sqlx-cli#force-building-in-offline-mode)
> 
> Setting `DATABASE_URL` in this `.env` does not work. I suspect the macro expander can't reach the database for some reason.

Gave this a try and still no code completion :/

**Comment by vlad20012** [comment]
@aldanor Most likely you need to update the IntelliJ Rust plugin

**Comment by longfellowone** [comment]
Receiving the following error `Access is denied. (os error 5)` #8238. When launching the app as an Administrator the error goes away. No alternative on mac

**Comment by longfellowone** [comment]
> I've noted that for sqlx I need to
> 
> * set `SQLX_OFFLINE=true` in my project's root `.env`
> * run `cargo sqlx prepare`
> 
> for `intellij-rust` to be able to expand `query!` macros. (both documented in https://github.com/launchbadge/sqlx/tree/ce801b933006c422ada464d0663c68dcf6c680d4/sqlx-cli#force-building-in-offline-mode)
> 
> Setting `DATABASE_URL` in this `.env` does not work. I suspect the macro expander can't reach the database for some reason.

Fixed: #8238 and #8236

**Comment by thurn** [comment]
Little confused about the status of #1786 -- the issue is shown with a checkmark here and a merged fix back in August, but the functionality doesn't work for me on the latest `0.4.162.4321-213` plugin build with `org.rust.macros.proc` enabled. Still waiting on something else?

**Comment by Undin** [comment]
@thurn could you provide an example where the plugin doesn't work with proc macros as expected, please?

**Comment by thurn** [comment]
reported as #8328

**Comment by longfellowone** [comment]
Having issues with proc macros again #8423

**Comment by hamza1311** [comment]
Is there any work going on for reporting errors generated by function-like macros? Right now, they just fail silently, forcing user to run `cargo check` to see what happened

**Comment by lfrancke** [comment]
I tried enabling this, refreshed a cargo project and it just worked.
Unfortunately, it only worked _once_. All changes since the initial refresh don't show up.
I tried refreshing, I disabled and reenabled the functionality but nothing worked.

My code compiles fine even though IntelliJ now shows it as "red".

Before I open a new ticket I thought I'd ask if I am maybe missing something: Do I need to do anything to manually update proc macro details and/or can I delete a cache somewhere? A `cargo clean` followed by a refresh works. Is this intended?

**Comment by johansmitsnl** [comment]
Had a issue with Sycamore (originally reported it for Yew) but these experimental features solved it for me. Also added screenshots and a repo to reproduce it: https://github.com/intellij-rust/intellij-rust/issues/6732#issuecomment-1248386202

**Comment by CristianKerr** [comment]
Hi I wanted to activate org.rust.macros.proc as an experimental feature, but I do not see it in there, what can I do to make it available? Thanks

**Comment by Undin** [comment]
@CristianKerr follow [this](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) instruction

**Comment by CristianKerr** [comment]
> @CristianKerr follow [this](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) instruction
Its not there
![Screenshot from 2022-10-21 12-17-30](https://user-images.githubusercontent.com/79404036/197173452-aa6cc8ac-7d91-4808-aa89-1bfea67ef2ce.png)

**Comment by Undin** [comment]
@CristianKerr I suppose you use **too** old plugin version

**Comment by CristianKerr** [comment]
> @CristianKerr I suppose you use **too** old plugin version

You are right, I have 0.3.140.3644-202, but I do not know why it doesn't show latest version for me (it looks like its up to date). Thanks!

**Comment by Undin** [comment]
> I do not know why it doesn't show latest version for me

Probably, you use old IDE version as well

**Comment by CristianKerr** [comment]
@Undin you are right, got stuck on older version, thanks for the help!

**Comment by Mersid** [comment]
I am not fully sure if this issue is related to procedural macros, but I'll post it here anyway. So, some functions from libraries do not show up in code completion:
![image](https://user-images.githubusercontent.com/30786211/207730664-a95bcd4c-e941-4b56-8ec1-9ba3d4ceae8f.png)

As one can see, on line 20 it shows that it is a perfectly valid function, but attempting to invoke the code completion dropdown will show that it cannot find a function by that name. I have all of the experimental features turned on, but it doesn't appear to help:
![image](https://user-images.githubusercontent.com/30786211/207730929-7b049545-51a8-449c-96b6-951fe59ee961.png)

Now, using the Re-Expand All Rust Macros command does fix it:
![image](https://user-images.githubusercontent.com/30786211/207731017-3ac8c88a-b1a8-400d-8892-00fb0f8ff2ec.png)

But it causes additional problems, because now CLion marks a lot of code as invalid and says things like "cannot move":
![image](https://user-images.githubusercontent.com/30786211/207731200-6d03caa7-ab83-4ca5-8d95-195c04796548.png)

Or "use of moved" value:
![image](https://user-images.githubusercontent.com/30786211/207731323-fa76c8f7-36f1-4972-a8d4-d2b036ea40e3.png)

The code does still work, and these errors did not show up before re-expanding macros. Additionally, re-expanding macros breaks the IDE's ability to detect and detect some macro-related code. Notice how BinaryTreeKey is in white; it appears that CLion no longer knows that BinaryTreeKey is a struct:
![image](https://user-images.githubusercontent.com/30786211/207731501-ef6f316e-8d70-4384-a592-570066639d8a.png)


---

Closing and re-opening the project does fix these issues:

![image](https://user-images.githubusercontent.com/30786211/207732066-ad186cac-aa9a-4eb3-b1db-48655f2c294d.png)
![image](https://user-images.githubusercontent.com/30786211/207732125-b3fdf9de-3c94-4b63-95f0-ed863836a34e.png)

But it also means that code completion can't find some functions again:
![image](https://user-images.githubusercontent.com/30786211/207732315-f4894503-d370-4318-a646-d3147a42083c.png)


Does anyone know how to fix this, or is this something that cannot be remedied?

Thanks!

**Comment by Mersid** [comment]
I think I found another issue. Macros that take structs don't show up properly if documentation in it ends with a backslash. For example, observe these two instances. They work fine normally
![image](https://user-images.githubusercontent.com/30786211/208342484-6523fa73-7821-43cc-873b-fcaa5a412dd8.png)
But when adding backslashes to the comment (and it only works with triple slash comments that end with a backslash), 
![image](https://user-images.githubusercontent.com/30786211/208342737-a6e87a07-d4a3-432e-9e32-924575d77f28.png)
The coloring gets stripped out, showing that the IDE is no longer able to properly recognize the code. I am not quite sure why this is happening, but it does mean that these types do not show up in code completion. An example is from the wgpu crate: the functions in wgpu::Features::* do not work because two of the bit flags end with a backslash in its comment.

**Comment by Mersid** [comment]
## Environment

* **IntelliJ Rust plugin version:** 0.4.185.5086-223
* **Rust toolchain version:** nightly-x86_64-pc-windows-msvc, rustc 1.68.0-nightly (37d7de337 2022-12-12)
* **IDE name and version:** CLion 2022.3
* **Operating system:** Windows 10

## Problem description
Derive macros do not appear to recognize procedural macros in code completion

## Steps to reproduce



1. Install something like bytemuck
2. Derive (for example) bytemuck::Zeroable
3. Observe that it should be valid, but if you enter bytemuck::Zeroabl (missing the last character) and hit Ctrl + Space to bring up the code completion, it will say that there are no suggestions.


For this test, ensure Cargo.toml has these two dependencies:
```toml
glam = {version = "0.22.0", features = ["bytemuck"]}
bytemuck = {version = "1.12.3", features = ["derive"]}
```

Create a struct:
```rust
#[repr(C)]
#[derive(Copy, Clone, Debug, bytemuck::Pod, bytemuck::Zeroabl)]
pub struct Vertex {
    position: glam::Vec3,
    color: glam::Vec3
}
```
Observe that:
![image](https://user-images.githubusercontent.com/30786211/208749604-8d243bb5-6215-4ef0-b05c-056046185977.png)

No suggestions. But if we were to manually enter it correctly?
![image](https://user-images.githubusercontent.com/30786211/208749714-5915bf62-405d-4c06-bd45-4156ffdd9cdc.png)
It recognizes it.

**Comment by djdisodo** [comment]
Span::call_site.source_file.path has empty path
